### PR TITLE
[Feature] TLS 1.3 0-RTT with replay protection

### PR DESCRIPTION
**Pull request by myrevery**
### Another premium feature to Hestia ###
Perhaps make Hestia the first to provide this protection by default.

How it works:

0rtt-anti-replay
### Attached ###
All templates are tidy and some fixes.

Prepare for upcoming HTTP/3 support.

### CODE REVIEW

**Comment by jaapmarcus** [comment]
Replacing https://github.com/hestiacp/hestiacp/pull/3686

**Comment by jaapmarcus** [comment]
@myrevery Is there any reason why this is still in draft?

**Comment by myrevery** [comment]
I just got up, only the upgrade script part, will be done soon...

**Comment by jaapmarcus** [comment]
No problem... :)

**Review by jaapmarcus — APPROVED** [review]

File: web/inc/main.php [hunk 1]
```diff
@@ -1,6 +1,12 @@
 <?php
 session_start();
 
+if ($_SERVER["HTTP_EARLY_DATA"] === "1") {
+	# Protection against early_data / 0-rrt
+	http_response_code(425);
+	exit();
+}
+
```
Comment (myrevery) [inline web/inc/main.php]: @jaapmarcus No need for this.

Seems you haven't fully understand how it works, don't hit me...:zany_face:

Potentially risky requests with early data never reach the backend.

Comment (myrevery) [reply web/inc/main.php]: I'll make some changes later to comply with RFC 8470, to avoid access failures caused by such application processing logic.

Comment (jaapmarcus) [reply web/inc/main.php]: I just did some testing:
https://internet.nl
And https://nabla-c0d3.github.io/sslyze/documentation/ and both still seem to report early_data enabled. Probably internet.nl uses sslyze as background. And sslyze uses an invalid request method that reports 400.

Also tested it with 
```
host=hestiacp.com # replace with your server name 
echo -e "HEAD / HTTP/1.1\r\nHost: $host\r\nConnection: close\r\n\r\n" > request.txt
openssl s_client -connect $host:443 -tls1_3 -sess_out session.pem -ign_eof < request.txt
openssl s_client -connect $host:443 -tls1_3 -sess_in session.pem -early_data request.txt
```
And the same issue but adding the php code solves the issue..

(Replaced hestiacp.com with my own domain)

Comment (myrevery) [reply web/inc/main.php]: This is the expected result.

I try to explain in detail how it works...

This mitigation is equivalent to taking over what the application is supposed to do.

We don't have an API to reject TLS handshake without modifying the NGINX source code.

So 0-RTT with valid PSK will be accepted (for TLS handshake only)

Then NGINX filters the HTTP request method:

idempotent or GET with no query will pass to the backend
non-idempotent or GET with query will return 30X/425

Valid client will automatically retry and then pass to the backend (the TLS handshake is done, so this time and subsequent requests will without early data)

And it doesn't add latency, just like 1-RTT.

Comment (myrevery) [reply web/inc/main.php]: If an attacker copies the 0-RTT request data (encrypted, with request method non-idempotent or GET with query)

Then attacker makes a replay attempt, but no matter how many times its replayed (when PSK is not expired)

NGINX will always return 30X/425 (encrypted)

No data will reach the backend.

**Comment by mattv8** [comment]
This is very cool, nice work @myrevery.

**Pull request merged**